applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -780,17 +780,25 @@ static void ble_qos_thread_fn(void)
 		u8_t *chmap;
 
 		chmap = chmap_filter_suggested_map_get(chmap_inst);
-		err = bt_le_set_chan_map(chmap);
-		if (err) {
-			LOG_WRN("bt_le_set_chan_map: %d", err);
-		} else {
-			LOG_DBG("Channel map update");
-			chmap_filter_suggested_map_confirm(chmap_inst);
 
-			k_mutex_lock(&data_access_mutex, K_FOREVER);
-			memcpy(current_chmap, chmap, sizeof(current_chmap));
-			k_mutex_unlock(&data_access_mutex);
+		struct ble_qos_event *event = new_ble_qos_event();
+		BUILD_ASSERT(sizeof(event->chmap) == CHMAP_BLE_BITMASK_SIZE, "");
+		memcpy(event->chmap, chmap, CHMAP_BLE_BITMASK_SIZE);
+		EVENT_SUBMIT(event);
+
+		if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
+			err = bt_le_set_chan_map(chmap);
+			if (err) {
+				LOG_WRN("bt_le_set_chan_map: %d", err);
+			} else {
+				LOG_DBG("Channel map update");
+			}
 		}
+
+		chmap_filter_suggested_map_confirm(chmap_inst);
```

Comment:
@nordic-auko, what do you think about it?

Reply:
I confirm the map as there is little point of not doing it. Regardless of what master will set we know that some channels should be excluded.
There is open question how the library retest channels that were excluded but I guess there must be some timeout inside the library itself (@nordic-auko correct me if I am wrong) that will reenable channels and allow to check crc errors there again.

I don't think we need to reflect exactly what controller sets. The idea of the QoS on the peripheral is to detect which channels are bad and this will be done.

Reply:
Hi, sorry I didn't have time to look at this yesterday. There will always be a time window where there is a mismatch between the actual channel map and the map known by the algorithm, as the controller doesn't tell you exactly when the new map will take effect.

This confirmation function is **not** tied to the channel rating logic, so if you omit calling this function the algorithm should work the same way. The reason for this confirmation function is such that the algorithm can determine if its internal continuously updated channel map recommendation is different from the actual map. If they are the same, there is no need to trigger a new channel map update.